server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -426,17 +429,66 @@ private static void addSortedBundle(Bundle bundle, Map<String, Bundle> bundles,
         Map<String, Plugin> loaded = new HashMap<>();
         Map<String, Set<URL>> transitiveUrls = new HashMap<>();
         List<Bundle> sortedBundles = sortBundles(bundles);
-
         for (Bundle bundle : sortedBundles) {
             checkBundleJarHell(JarHell.parseClassPath(), bundle, transitiveUrls);
 
             final Plugin plugin = loadBundle(bundle, loaded);
             plugins.add(new Tuple<>(bundle.plugin, plugin));
         }
 
+        loadExtensions(plugins);
         return Collections.unmodifiableList(plugins);
     }
 
+    // package-private for test visibility
+    static void loadExtensions(List<Tuple<PluginInfo, Plugin>> plugins) {
+        Map<String, List<Plugin>> extendingPlugins = plugins.stream()
+            .flatMap(t -> t.v1().getExtendedPlugins().stream().map(extendedPlugin -> Tuple.tuple(extendedPlugin, t.v2())))
+            .collect(Collectors.groupingBy(Tuple::v1, Collectors.mapping(Tuple::v2, Collectors.toList())));
+        plugins.stream().filter(t -> t.v2() instanceof ExtensiblePlugin).forEach(t -> loadExtensionsForPlugin((ExtensiblePlugin) t.v2(),
+            extendingPlugins.getOrDefault(t.v1().getName(), List.of())));
+    }
+
+    private static void loadExtensionsForPlugin(ExtensiblePlugin extensiblePlugin, List<Plugin> extendingPlugins) {
+        ExtensiblePlugin.ExtensionLoader extensionLoader = new ExtensiblePlugin.ExtensionLoader() {
```

Comment:
This could just be created in loadExtensions, no need to create a new one for each individual plugin? AFAICT it always operates on the same stream of plugins passed in.

Reply:
The set of plugins passed in are only those extending the plugin that we load extensions for. This ensures that a plugin cannot ask for extensions that it does not own. Like ML asking for all autoscaling-deciders, which would succeed if the list of plugins was every plugin extending some other plugins. Also, the runtime would be slightly worse I think, since we would also be searching for extensions in unrelated modules.

Reply:
Thanks for the explanation. I missed the lookup in calling of this method, thinking the extendingPlugins in the calling function (a map) was the same as extendingPlugins here (the value from that map, a list). Maybe consider changing the variable name to clarify they are different.